tools/free_space.awk

```diff
@@ -1,13 +1,74 @@
-#!/usr/bin/awk -f
+#!/usr/bin/gawk -f
 
-# Usage: tools/free_space.awk pokecrystal.map
+# Usage:  tools/free_space.awk [BANK=<bank_spec>] pokecrystal.map
+
+# Tha BANK argument allows printing free space in one, all, or none of the ROM's banks
+# Valid arguments are numbers (in decimal "42" or hexadecimal "0x2a"), "all" or "none".
+# If not specified, defaults to "none".
+# The `BANK` argument MUST be before the map file name, otherwise it has no effect!
+# Yes:  tools/free_space.awk BANK=all pokecrsyal.map
+# No:   tools/free_space.awk pokecrysal.map BANK=42
+
+# Copyright (c) 2020, Eldred Habert.
+# SPDX-License-Identifier: MIT
 
 BEGIN {
-	total = free = 16384 * 128
+	nb_banks = 0
+	free = 0
+	rom_bank = 0 # Safety net for malformed files
+
+	# if BANK is a number, convert it to one
+	if (BANK ~ /([0-9]+|0[xX][0-9a-fA-F]+)/) {
+		BANK = strtonum("none")
+	} else {
+		# Otherwise, preserve "all" and turn everything else into an impossibility
+		BANK ~ /all/ || BANK = -0.5 # We only deal with positive integers, let's be extra safe
+	}
 }
-/^  SECTION: \$[0-7]/ {
-	free -= strtonum("0x" substr($3, 3))
+
+# Only accept ROM banks, ignore everything else
+toupper($0) ~ /^[ \t]*ROM[0X][ \t]+BANK[ \t]+#/ {
+	nb_banks++
+	rom_bank = 1
+	split($0, fields, /[ \t]/)
+	bank_num = strtonum(substr(fields[3], 2))
+}
+
+function register_bank(amount) {
+	free += amount
+	rom_bank = 0 # Reject upcoming banks by default
+
+	if (BANK ~ /all/ || BANK == bank_num) {
+		printf "Bank %3d: %5d/16384 (%.2f%%)\n", bank_num, amount, amount * 100 / 16384
+	}
 }
+
+rom_bank && toupper($0) ~ /^[ \t]*EMPTY([ \t]|$)/ {
+	# Empty bank
+	register_bank(16384)
+}
+rom_bank && toupper($0) ~ /^[ \t]*SLACK:[ \t]/ {
+	if ($2 ~ /\$[0-9A-F]+/) {
+		register_bank(strtonum("0x" substr($2, 2)))
+	} else {
+		printf "Malformed slack line? \"%s\" does not start with '$'\n", $2
+	}
+}
+
 END {
-	printf "Free space: %d/%d (%.2f%%)\n", free, total, free * 100 / total
+	# Determine number of banks, by rounding to upper power of 2
+	total_banks = 2 # Smallest size is 2 banks
```

Comment:
Yeah, I've optimized one project enough that it has >50% free space and builds a 1MB ROM, and other times when it's gone over 2MB it would be nice to see negative free space.

Reply:
> Isn't the size of the ROM hardcoded through the related rgbfix flag?

[Nope, it rounds up.](http://rednex.github.io/rgbds/rgbfix.1.html#p)

The goal of that modification is to work even with hacks that change the ROM size, even if they use a different ROM size.

Reply:
It's more likely that a hack just got too large or small than that it changed the valid ROM size.

https://github.com/Rangi42/rgbs builds a 1 MB ROM; this new script prints `Free space: 360717/1048576 (34.40%)` even though it should be over 50%.

And without modifying the MBC settings, rgbds can still overflow content into banks $80 and above and output a 4MB ROM. I've had extra music data end up there and either not play at all in BGB, or cause a crash. It shouldn't claim I have nearly 50% free space just because bank $80 is slightly full.

I think a comment explaining that a hard-coded bank would need updating if you change the MBC/ROM size would suffice.

Reply:
Then it's not the script's fault, the ROM is broken. If you have such a problem, you have a different issue entirely. Knowledge of the max size acceptable by the different MBCs is something the programmer should have anyways.

Reply:
That's not a broken ROM though? A 1 MB ROM is what rgbds 0.4.0 outputs for pokecrystal when there aren't enough SECTIONs to cover more than $40 banks. If it needs to be 2 MB when using MBC3, then rgbds should be fixed to always output that (or we should be specifying a fixed cartridge size in pokecrystal's Makefile, not this auto-pad value). As for the 4MB MBC3 ROMs with too much content, yes, that's a problem, but the point of running a free-space script is to help notice that problem. (And again, is a problem that a fixed cart size in the Makefile could avoid.)